src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Unix.cs

```diff
@@ -61,5 +61,97 @@ internal static unsafe void GetAnsiStringBytes(ReadOnlySpan<char> chars, Span<by
             int actualByteLength = Encoding.UTF8.GetBytes(chars, bytes);
             bytes[actualByteLength] = 0;
         }
+
+        public static IntPtr AllocHGlobal(IntPtr cb)
+        {
+            nuint cbNative = (nuint)(nint)cb;
+
+            // Avoid undefined malloc behavior by always allocating at least one byte
+            IntPtr pNewMem = Interop.Sys.MemAlloc((cbNative != 0) ? cbNative : 1);
+            if (pNewMem == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            return pNewMem;
+        }
+
+        public static void FreeHGlobal(IntPtr hglobal)
+        {
+            if (hglobal != IntPtr.Zero)
+            {
+                Interop.Sys.MemFree(hglobal);
+            }
+        }
+
+        public static IntPtr ReAllocHGlobal(IntPtr pv, IntPtr cb)
+        {
+            nuint cbNative = (nuint)(nint)cb;
+
+            // Match Windows behavior by always allocating at least one byte and only when the original pointer is non-zero
+            IntPtr pNewMem = (pv != IntPtr.Zero) ? Interop.Sys.MemReAlloc(pv, (cbNative != 0) ? cbNative : 1) : IntPtr.Zero;
+            if (pNewMem == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+            return pNewMem;
+        }
+
+        public static IntPtr AllocCoTaskMem(int cb) => AllocHGlobal((nint)(uint)cb);
+
+        public static void FreeCoTaskMem(IntPtr ptr) => FreeHGlobal(ptr);
+
+        public static IntPtr ReAllocCoTaskMem(IntPtr pv, int cb)
+        {
+            nuint cbNative = (nuint)(uint)cb;
+
+            if (cbNative == 0)
+            {
+                if (pv != IntPtr.Zero)
+                {
+                    Interop.Sys.MemFree(pv);
+                    return IntPtr.Zero;
+                }
+                // Avoid undefined realloc behavior by always allocating at least one byte
+                cbNative = 1;
+            }
+
+            IntPtr pNewMem = Interop.Sys.MemReAlloc(pv, cbNative);
+            if (pNewMem == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+            return pNewMem;
+        }
+
+        internal static unsafe IntPtr AllocBSTR(int length)
+        {
+            const nuint alignment = 16;
```

Comment:
nit: `WIN32_ALLOC_ALIGN` is defined as `16 - 1` in CoreCLR. Should this match (to avoid writing `alignment -1` twice)?

Reply:
Fixed. I have also kept the original constant name so that it is easy to find.